src/dataset/impl_dataset.rs

```diff
@@ -724,50 +740,119 @@ where
+            let mut targets_sl2 = self.targets.as_multi_targets_mut();
+            let targets_sl = targets_sl2.as_slice_mut().unwrap();
+
+            for i in 0..k {
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+
+                {
+                    let train = DatasetBase::new(
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, features),
+                            records_sl.split_at(fold_size * features).1,
+                        )
+                        .unwrap(),
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, targets),
+                            targets_sl.split_at(fold_size * targets).1,
+                        )
+                        .unwrap(),
+                    );
+
+                    let obj = fit_closure(&train);
+                    objs.push(obj);
+                }
+
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+            }
         }
 
         objs.into_iter().zip(self.sample_chunks(fold_size))
-        //
     }
-}
 
-fn assist_swap_array2<F>(slice: &mut [F], index: usize, fold_size: usize, features: usize) {
-    if index == 0 {
-        return;
+    pub fn cross_validate_mt<O, DT, ER: std::error::Error, M, C>(
+        &'a mut self,
+        k: usize,
+        parameters: &[M],
+        eval: C,
+    ) -> Array1<F>
+    where
+        DT: Data<Elem = E>,
+        M: for<'c> Fit<ArrayView2<'c, F>, ArrayView2<'c, E>, ER, Object = O>,
+        O: for<'d> PredictRef<ArrayView2<'a, F>, ArrayBase<DT, Ix2>>,
+        C: Fn(&ArrayBase<DT, Ix2>, &ArrayView2<E>) -> F,
+    {
+        let mut accuracies = Array1::from_elem(parameters.len(), F::zero());
+        for acc in self
+            .iter_fold(k, |train| {
+                parameters
+                    .iter()
+                    .map(|p| p.fit(&train).unwrap())
+                    .collect::<Vec<_>>()
+            })
+            .map(|(models, valid)| {
+                models
+                    .iter()
+                    .map(|m| {
+                        let predicted = m.predict(valid.records());
+                        eval(&predicted, valid.targets())
+                    })
+                    .collect::<Array1<F>>()
+            })
+        {
+            for i in 0..accuracies.len() {
+                accuracies[i] += acc[i];
+            }
+        }
+        accuracies
+    }
+
+    pub fn cross_validate<O, DT, ER: std::error::Error, M, C>(
+        &'a mut self,
+        k: usize,
+        parameters: &[M],
+        eval: C,
+    ) -> Array1<F>
+    where
+        DT: Data<Elem = E>,
+        M: for<'c> Fit<ArrayView2<'c, F>, ArrayView2<'c, E>, ER, Object = O>,
```

Comment:
can't you make this generic with 
```
M: for<'c> Fit<ArrayView2<'c, F>,T, ER, Object = O>,
T: AsTargets<Elem = E>
```
for multi and single targets?

Reply:
The problem is the prediction. Single target algorithms do not have an implementation of `PredictRef` returning an Array2 and so things can get a little tricky when giving arguments to the closure. I'll see if I can come up with something, but it would mean adding a match on the number of dimensions of the input targets, which could be two because it is an actual multi-target model or just because an axis was added. Then if it it the latter case the user would get a multi-target evaluation (like `r2`) and would have to call `[0]` on it which is not very pretty

Reply:
have you tried to add a `I: Dimension` bound and then make `O` and `C` generic over that dimensionality? Something like:
```rust
O: for<'d> PredictRef<ArrayView2<'a, F>, ArrayBase<DT, I>>,
C: Fn(&ArrayBase<DT, I>, &ArrayBase<DT, I>) -> F,
I: Dimension
```

Reply:
I guess that my main point is that I can't distinguish between multi-target datasets and single-target datasets if not by checking the dimension of the second axis at runtime, and I wanted to avoid forcing the user to average over 1 value, but the only way to do so is to call `try_single_target` in a dedicated function. Or at least this is what i concluded earlier, tomorrow I'll play with it and see if I can figure out something.

Reply:
Basically, you can have just one of the following two nice properties:

* Having just one function `cross_validate` which handles both single and multi-dimensional cases
* Having the ability to call `try_single_target` on "fake" multi-dimensional datasets

Knowing that all functions in the `dataset` crate return an `Array2` as targets by default.

Basically, I can parameterize by dimension as you mention but then I cannot call `try_single_target` (inferring that the dataset should have single targets from the predict method signature), so the user would have to do something like this:

```rust
let mut dataset = linfa_datasets::diabetes();

let models = vec![whatever];

let evaluations = dataset.cross_validate(5, &models, |prediction, truth| truth.r2(prediction).unwrap()[0]);
```

because regression metrics automatically compute multi-target metrics if the caller is an array2. This could get even more confusing since `prediction.r2(truth)` is single-target if the prediction is too.

Also, I would like the `eval` closure to return `F` in single-dimensional cases and `Array1<F>` in multi-dimensional cases, so that one can check the fold-averaged evaluation for every target separately and then perform averaging or whatever else they want, without having to average across targets in every fold, and I don't know how I can do that with generics.

Reply:
Also, maybe the eval closure should return a result so that `r2?` and `confusion_matrix()?.accuracy()` can be used